cpp/util/ManagedArray.h

```diff
@@ -0,0 +1,146 @@
+#ifndef NUMERICAL_ARRAY_H
+#define NUMERICAL_ARRAY_H
+
+#include <memory>
+#include <tbb/tbb.h>
+
+/*! \file ManagedArray.h
+    \brief Defines the standard array class to be used throughout freud.
+*/
+
+namespace freud { namespace util {
+
+//! Class to handle the storage of all arrays of numerical data used in freud.
+/*! The purpose of this class is to handle standard memory management, and to
+ *  provide an abstraction around the implementation-specific choice of
+ *  underlying data structure for arrays of data in freud. These arrays are
+ *  specifically designed for numerical data calculated by a compute class,
+ *  particularly for arrays that must be made accessible through the Python API
+ *  of freud.
+ *
+ *  To support resizing, a ManagedArray instances stores its data as a pointer
+ *  to a pointer, and its size as a pointer. As a result, copy-assignment or
+ *  initialization will result in a new ManagedArray pointing to the same data,
+ *  and any such array can resize or reallocate this data. The pointer to
+ *  pointer infrastructure ensures that such changes properly propagate to all
+ *  ManagedArrays referencing a given memory space. In addition, this
+ *  infrastructure allows the creation of a completely new ManagedArray with a
+ *  new set of pointers that also manages the same data, allowing it to keep
+ *  the original array alive if the original ManagedArray instances become
+ *  decoupled from it.
+ */
+template<typename T> class ManagedArray
+{
+public:
+    //! Default constructor with optional size for thread local arrays
+    /*! \param size Size of the array to allocate.
+     */
+    ManagedArray(unsigned int size=0)
+    {
+        m_size = std::make_shared<unsigned int>(size);
+        // Hard code making an array of at least size 1.
+        m_data = std::shared_ptr<std::shared_ptr<T> >(
```

Comment:
This seems very odd, can you help me understand why another allocation is necessary?

Reply:
This is the same kludge as above.